ibm/service/power/data_source_ibm_pi_network_security_group.go

```diff
@@ -0,0 +1,314 @@
+													Computed:    true,
+													Description: "TCP flag.",
+													Type:        schema.TypeString,
+												},
+											},
+										},
+										Type: schema.TypeList,
+									},
+									Attr_Type: {
+										Computed:    true,
+										Description: "The protocol of the network traffic.",
+										Type:        schema.TypeString,
+									},
+								},
+							},
+							Type: schema.TypeList,
+						},
+						Attr_Remote: {
+							Computed:    true,
+							Description: "List of remote.",
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									Attr_ID: {
+										Computed:    true,
+										Description: "The ID of the remote Network Address Group or network security group the rules apply to. Not required for default-network-address-group.",
+										Type:        schema.TypeString,
+									},
+									Attr_Type: {
+										Computed:    true,
+										Description: "The type of remote group the rules apply to.",
+										Type:        schema.TypeString,
+									},
+								},
+							},
+							Type: schema.TypeList,
+						},
+						Attr_SourcePort: {
+							Computed:    true,
+							Description: "ist of source port",
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									Attr_Maximum: {
+										Computed:    true,
+										Description: "The end of the port range, if applicable, If values are not present then all ports are in the range.",
+										Type:        schema.TypeInt,
+									},
+									Attr_Minimum: {
+										Computed:    true,
+										Description: "The start of the port range, if applicable. If values are not present then all ports are in the range.",
+										Type:        schema.TypeInt,
+									},
+								},
+							},
+							Type: schema.TypeList,
+						},
+					},
+				},
+				Type: schema.TypeList,
+			},
+			Attr_UserTags: {
+				Computed:    true,
+				Description: "The user tags associated with this resource.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Type:        schema.TypeList,
+			},
+		},
+	}
+}
+
+func dataSourceIBMPINetworkSecurityGroupRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := d.Get(Arg_CloudInstanceID).(string)
+	nsgClient := instance.NewIBMIPINetworkSecurityGroupClient(ctx, sess, cloudInstanceID)
+
+	networkSecurityGroup, err := nsgClient.Get(d.Get(Arg_NetworkSecurityGroupID).(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(*networkSecurityGroup.ID)
+	d.Set(Attr_CRN, networkSecurityGroup.Crn)
+
+	if len(networkSecurityGroup.Members) > 0 {
+		members := []map[string]interface{}{}
+		for _, mbr := range networkSecurityGroup.Members {
+			mbrMap := networkSecurityGroupMemberToMap(mbr)
+			members = append(members, mbrMap)
+		}
+		d.Set(Attr_Members, members)
+	}
```

Comment:
did you not have to add to stop a double apply situation? I ran into that with rule-add code
```
else {
		d.Set(Attr_Members, nil)
	}
```

Reply:
Yes, I have added `nil` for empty `rules` and `members`